### PR TITLE
Exercise 8

### PR DESCRIPTION
**Pull request by maheshpatelhome**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,5 +1,23 @@
...
+services: docker
+
+jobs: 
+  include:
+    - stage: test
+      script:
+      - docker build --target test --tag my-test-image .
+      - docker run my-test-image ./todo_app/tests/todo_view_model_test.py
+      - docker run my-test-image ./todo_app/tests/test_client.py
+      - docker run -e API_KEY -e API_TOKEN -e BOARD_NAME -e MEMBERS_NAME -e TO_DO_LIST_ID -e DOING_LIST_ID -e DONE_LIST_ID  -e FLASK_APP -e FLASK_ENV -e SECRET_KEY my-test-image  todo_app/tests_e2e
```
Comment (hugh-emerson) [inline .travis.yml]: MEMBERS_NAME is not used by your app at all. 

And BOARD_NAME, TODO_LIST_ID, DOING_LIST_ID, DONE_LIST_ID are not used by the e2e test - it creates its own temporary board.

File: .travis.yml [hunk 2]
```diff
@@ -1,5 +1,23 @@
...
+    - stage: deploy
+      if: branch = main
```
Comment (hugh-emerson) [inline .travis.yml]: When a pull request triggers a build, "branch" is equal to the target branch. That means you are currently deploying whenever main is updated, **or** someone opens a pull request to "main", which is not what we want. You should only deploy after it is merged.

File: .travis.yml [hunk 3]
```diff
@@ -1,5 +1,23 @@
...
+      - docker build --target test --tag my-test-image .
+      - docker run my-test-image ./todo_app/tests/todo_view_model_test.py
+      - docker run my-test-image ./todo_app/tests/test_client.py
+      - docker run -e API_KEY -e API_TOKEN -e BOARD_NAME -e MEMBERS_NAME -e TO_DO_LIST_ID -e DOING_LIST_ID -e DONE_LIST_ID  -e FLASK_APP -e FLASK_ENV -e SECRET_KEY my-test-image  todo_app/tests_e2e
+    - stage: deploy
+      if: branch = main 
+      script:
+      - echo $DOCKER_PASSWORD | docker login --username "$DOCKER_USERNAME" --password-stdin
+      - docker build --target production --tag "$DOCKER_USERNAME"/todo-app-prod:"$TRAVIS_COMMIT" .
+      - docker push "$DOCKER_USERNAME"/todo-app-prod:"$TRAVIS_COMMIT"
```
Comment (hugh-emerson) [inline .travis.yml:19]: This is nice but you would probably want to tag/push a `:latest` image as well.

**Review by hugh-emerson — APPROVED** [review]
Looks good 👍 
Some minor comments but merge when you're ready

File: .travis.yml (hunk 1 above)
Comment (maheshpatelhome) [reply .travis.yml]: Removed unneccesary env variables

File: .travis.yml (hunk 2 above)
Comment (maheshpatelhome) [reply .travis.yml]: added code to check for a PUSH on the main branch before deploying

File: .travis.yml (hunk 3 above)
Comment (maheshpatelhome) [reply .travis.yml:19]: done

**Pull request merged**